_posts/2021-03-12-jdk16-g1-parallel-gc-changes.md

```diff
@@ -6,53 +6,57 @@ date:   2021-03-12 16:39:24 +0200
 
-  * Similarly there has been an improvement to **determine the heap occupancy threshold when to start this marking** in [JDK-8245511](https://bugs.openjdk.java.net/browse/JDK-8245511). This heap occupancy threshold is computed using recent allocation rate in the old generation and how long it typically takes to start the first old generation garbage collection from that start of the marking. The larger the allocation rate and the longer the time from the start of marking to the first collection in the old generation, the lower this threshold is, i.e. the earlier G1 starts the marking.
+  * Similarly there has been an improvement to **determine the heap occupancy threshold when to start this marking** in [JDK-8245511](https://bugs.openjdk.java.net/browse/JDK-8245511). This heap occupancy threshold is computed using recent allocation rate in the old generation and how long it typically takes to start the first old generation garbage collection from the start of the marking. The larger the allocation rate and the longer the time from the start of marking to the first collection in the old generation are, the lower this threshold is, i.e. the earlier G1 starts the marking.
+  
+    **OR** The larger the allocation rate is and/or the longer the time from the start of marking to the first collection in the old generation is, the lower this threshold is, i.e. the earlier G1 starts the marking.
 
-    The problem are humongous objects that are always directly allocated in the old generation, so they typically factor heavily into this allocation rate. However, since JDK 9, with the eager reclamation mentioned above, these humongous objects can reclaimed very quickly and the overall allocation rate in the old generation is actually very low. This led to many unnecessary concurrent markings. The mentioned change for JDK-8245511 solves this problem by taking eager reclamation into account better. There are some nice graphs showing the impact in the [bug report](https://bugs.openjdk.java.net/browse/JDK-8245511).
+    Humongous objects are (were?) problematic as they are always directly allocated in the old generation, so they typically factor heavily into this allocation rate. However, since JDK 9, with the eager reclamation mentioned above, these humongous objects can reclaimed very quickly and the overall allocation rate in the old generation is actually very low. This led to many unnecessary concurrent markings. The mentioned change for JDK-8245511 solves this problem by taking eager reclamation better into account. There are some nice graphs showing the impact in the [bug report](https://bugs.openjdk.java.net/browse/JDK-8245511).
+	
+	(Why don't you include the graph here? post with visual will alwats attracts more reads).
```

Comment:
The graphs are very hard to read, would need extensive explanation, and would need re-taking of the measurements to get them into a style so that I would be satisfied with them too. They are certainly fine for the bug report and to get a development discussion going on, and "nice" from a results POV.

This post is also mostly meant as some kind of signpost for the audience, not full explanation of the changes. I do try to spend a few sentences each to explain them, the reasons why they were proposed and their impact, but there is a limit on the amount of effort I want to spend on each.

Ideally there would be a post to link to already.